crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -304,110 +296,327 @@ where
+            validate_transparent_input(
+                vin,
+                changed_balances,
+                transparent_tx_pool,
+                epoch,
+                conversion_state,
+                signers,
+            )?;
+        }
+
+        for out in transp_bundle.vout.iter() {
+            validate_transparent_output(
+                out,
+                changed_balances,
+                transparent_tx_pool,
+                epoch,
+                conversion_state,
+            )?;
+        }
+    }
+    Ok(())
+}
+
+// Verify that the pre balance + the Sapling value balance = the post balance
+// using the decodings in tokens and conversion_state for assistance.
+fn verify_sapling_balancing_value(
+    pre: &ValueSum<Address, Amount>,
+    post: &ValueSum<Address, Amount>,
+    sapling_value_balance: &I128Sum,
+    target_epoch: Epoch,
+    tokens: &BTreeMap<AssetType, (Address, token::Denomination, MaspDigitPos)>,
+    conversion_state: &ConversionState,
+) -> Result<()> {
+    let mut acc = ValueSum::<Address, SignedAmount>::from_sum(pre.clone());
+    for (asset_type, val) in sapling_value_balance.components() {
+        // Only assets with at most the target timestamp count
+        match conversion_state.assets.get(asset_type) {
+            Some(((address, _, digit), asset_epoch, _, _))
+                if *asset_epoch <= target_epoch =>
+            {
+                let decoded_change = SignedAmount::from_masp_denominated(
+                    *val, *digit,
+                )
+                .map_err(|_| {
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Overflow in MASP value balance",
+                    ))
+                })?;
+                let decoded_change =
+                    ValueSum::from_pair(address.clone(), decoded_change);
+                acc = acc.checked_sub(&decoded_change).ok_or_else(|| {
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Overflow in MASP value balance",
+                    ))
+                })?;
+            }
+            None if tokens.contains_key(asset_type) => {
+                let (token, _denom, digit) = &tokens[asset_type];
+                let decoded_change = SignedAmount::from_masp_denominated(
+                    *val, *digit,
+                )
+                .map_err(|_| {
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Overflow in MASP value balance",
+                    ))
+                })?;
+                let decoded_change =
+                    ValueSum::from_pair(token.clone(), decoded_change);
+                acc = acc.checked_sub(&decoded_change).ok_or_else(|| {
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Overflow in MASP value balance",
+                    ))
+                })?;
```

Comment:
this code looks nearly identical to me. can it be abstracted into a method/fn?

Reply:
Thanks, good point. Done.